### PR TITLE
Ensemble parallel

### PR DESCRIPTION
**Pull request by trbromley**
This tutorial shows:

- How to use two QPUs in an ensemble model on iris dataset
- That the ensemble is better than the individual models
- An example of using Rigetti and Qiskit devices together

### CODE REVIEW

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+            markersize=10,
+        ),
+    ]
+
+    ax.legend(handles=custom_lines, bbox_to_anchor=(1.0, 0.75))
+
+    return ax
+
+
+plot_points(x_train, y_train, x_test, y_test)
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: Any way to suppress the matplotlib text output?
e.g., I'm seeing `<matplotlib.axes._subplots.AxesSubplot object at 0x7f6d5b6988d0>` in addition to the figure

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+#
+# Quantum nodes
+# ^^^^^^^^^^^^^
+#
+# We begin by defining the two quantum devices and the circuits to be run on them.
+
+
+n_wires = 4
+
+# Use forest.qpu if hardware access is available
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
# You can use "forest.qpu" if hardware access is available
```

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+dev0 = qml.device("forest.qvm", device="Aspen-4-4Q-E")
+
+# Use qiskit.ibmq if access is available
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
# You can use "qiskit.ibmq" if hardware access is available
```
Is this the right way to specify HW device?

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+dev1 = qml.device("qiskit.aer", wires=4)
+devs = [dev0, dev1]
+
+##############################################################################
+# The circuits run on both devices are shown in the figure below:
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
# The circuits for both QPUs are shown in the figure below:
```

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    qml.CZ(wires=[1, 2])
+    qml.CZ(wires=[1, 3])
+
+    for i in range(n_wires):
+        qml.Rot(*params[0, 1, i], wires=i)
+    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1)), qml.expval(qml.PauliZ(2))
+
+
+##############################################################################
+# We finally combine the two devices into a :class:`~.pennylane.QNodeCollection`:
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
# We finally combine the two devices into a :class:`~.pennylane.QNodeCollection` that uses the PyTorch interface:
```

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+##############################################################################
+# Postprocessing into a prediction
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# The ``predict_point`` function below allows us to find the ensemble prediction, as well as keeping
+# track of the individual predictions from each QPU.
+#
+# We include a ``parallel`` keyword argument for evaluating the :class:`~.pennylane.QNodeCollection`
+# in a parallel asynchronous manner. This feature requires the ``dask`` library, which can be
+# installed using ``dask[delayed]``. When ``parallel=True``, we are able to make
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: I had trouble installing dask[delayed]. Perhaps we should give a more explicit command or point to the dask installation page?

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+# Quantum nodes
+# ^^^^^^^^^^^^^
+#
+# We begin by defining the two quantum devices and the circuits to be run on them.
+
+
+n_wires = 4
+
+# Use forest.qpu if hardware access is available
+dev0 = qml.device("forest.qvm", device="Aspen-4-4Q-E")
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: We should have a callout that notifies the user that if they want to run this code locally, they will have to make sure the qvm (and quilc) need to be running in the background (perhaps providing a link to forest sdk instructions).

Note that the full page does not run correctly unless both qvm and quilc are turned on

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+# These numbers tell us a few things:
+#
+# - On both training and test datasets, the ensemble model outperforms the predictions from each
+#   QPU. This provides a nice example of how QPUs can be used in parallel to gain a performance
+#   advantage.
+#
+# - The accuracy of QPU0 is much higher than the accuracy of QPU1. This does not mean that one
+#   device is intrinsically better than the other. In fact, another set of parameters can lead to
+#   QPU1 becoming more accurate. We see in the following that the difference in accuracy is due
+#   to specialization of each QPU.
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
#   to specialization of each QPU, which leads to overall better performance of the ensemble model.
```

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+##############################################################################
+# These numbers tell us a few things:
+#
+# - On both training and test datasets, the ensemble model outperforms the predictions from each
+#   QPU. This provides a nice example of how QPUs can be used in parallel to gain a performance
+#   advantage.
+#
+# - The accuracy of QPU0 is much higher than the accuracy of QPU1. This does not mean that one
+#   device is intrinsically better than the other. In fact, another set of parameters can lead to
+#   QPU1 becoming more accurate. We see in the following that the difference in accuracy is due
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
#   QPU1 becoming more accurate. We will see in the next section that the difference in accuracy is due
```

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+#   performance of the ensemble model, rather than minimizing the generalization error.
+#
+# Choice of QPU
+# ^^^^^^^^^^^^^
+#
+# Is there a link between the class of a datapoint and the QPU chosen to make the prediction in
+# the ensemble model? Let's investigate.
+
+
+# Combine choices data to simplify analysis
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: While reading this, it was unclear what "choices" meant :thinking:

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+# We can hence find the predictions each QPU was responsible for.
+
+
+choices_vs_prediction_0 = choice_vs_prediction[choice_vs_prediction[:, 0] == 0]
+choices_vs_prediction_1 = choice_vs_prediction[choice_vs_prediction[:, 0] == 1]
+predictions_0 = choices_vs_prediction_0[:, 1]
+predictions_1 = choices_vs_prediction_1[:, 1]
+
+
+expl = "When the QPU{} was chosen by the ensemble, it made the following distribution of " \
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: Text is slightly too long (a few characters), leading to a horizontal scrollbar, which we don't want

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+    ax.legend(handles=custom_lines, bbox_to_anchor=(1.0, 0.75))
+
+    return ax
+
+
+##############################################################################
+# We can again compare the ensemble model with the individual models from each QPU.
+
+
+plot_points_prediction(x, y, predictions, "ensemble")  # ensemble
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: again, matplotlib is outputing an unnecessary line of text for each figure

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+
+plot_points_prediction(x, y, np.append(p_train_0, p_test_0), "QPU0")  # QPU 0
+
+##############################################################################
+
+plot_points_prediction(x, y, np.append(p_train_1, p_test_1), "QPU1")  # QPU 1
+
+##############################################################################
+# These plots reinforce the specialization of the two QPUs. QPU1 concentrates on doing a good job
+# at predicting class 1 points, while QPU0 is focused on classes 0 and 2. By combining together,
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]:
```suggestion
# at predicting class 1, while QPU0 is focused on classes 0 and 2. By combining together,
```

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+# the resultant ensemble performs better.
+#
+# This tutorial shows how QPUs can work in parallel to realize a performance advantage. Check out
+# our :doc:`../app/tutorial_vqe_parallel.py` tutorial to see how multiple QPUs can be
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: reference does not resolve properly

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+import numpy as np
+import pennylane as qml
+import sklearn.datasets
+import sklearn.decomposition
+import torch
+from matplotlib.lines import Line2D
+from matplotlib.patches import Patch
+
+##############################################################################
+# This tutorial requires the ``pennylane-forest`` and ``pennylane-qiskit`` packages, which can be
```
[the remark this answers is not in the document]
Comment (josh146) [reply implementations/tutorial_ensemble_multi_qpu.py]: (note: best to link to the pytorch getting started page for that, since the pip install version depends on os/python version)

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+
+data = sklearn.datasets.load_iris()
+x = data["data"]
+y = data["target"]
+
+##############################################################################
+# We shuffle the data and then embed the four features into a two-dimensional space for ease of
+# plotting later on. The first two components of a PCA are used.
+
+np.random.seed(1967)
```
[the remark this answers is not in the document]
Comment (josh146) [reply implementations/tutorial_ensemble_multi_qpu.py]: You can tell who wrote the tutorial solely based on seed values :laughing:

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+
+
+n_wires = 4
+
+# Use forest.qpu if hardware access is available
+dev0 = qml.device("forest.qvm", device="Aspen-4-4Q-E")
+
+# Use qiskit.ibmq if access is available
+dev1 = qml.device("qiskit.aer", wires=4)
+devs = [dev0, dev1]
```
[the remark this answers is not in the document]
Comment (josh146) [reply implementations/tutorial_ensemble_multi_qpu.py]: Yep; if you have docker installed, you can run this:
```
docker run --rm -it rigetti/quilc -R
```

File: implementations/tutorial_ensemble_multi_qpu.py (hunk 3 above)
Comment (josh146) [reply implementations/tutorial_ensemble_multi_qpu.py]: You would also have to specify the backend:
```python
dev = qml.device("qiskit.ibmq", backend="ibmqx5")
```

File: implementations/tutorial_ensemble_multi_qpu.py (hunk 6 above)
Comment (josh146) [reply implementations/tutorial_ensemble_multi_qpu.py]: Link is more maintainable :+1:

File: implementations/tutorial_ensemble_multi_qpu.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply implementations/tutorial_ensemble_multi_qpu.py]: I'm thinking more about users running this code locally (i.e., people who won't have docker), and they might be confused if they get this error

[the remark this answers is not in the document]
Comment (josh146) [reply implementations/tutorial_ensemble_multi_qpu.py]: Oh, we should link to wherever the latest version of the Forest installation instructions are.

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 18]
```diff
@@ -108,30 +111,14 @@ def plot_points(x_train, y_train, x_test, y_test):
         Patch(facecolor=colours[0], edgecolor=c_transparent, label="Class 0"),
         Patch(facecolor=colours[1], edgecolor=c_transparent, label="Class 1"),
         Patch(facecolor=colours[2], edgecolor=c_transparent, label="Class 2"),
-        Line2D(
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: Doing it this way causes a horizontal scrollbar to appear. Can we go halfway, e.g., by doing a line break at `markerfacecolor` (same applies to both `Line2D`s)

File: implementations/tutorial_ensemble_multi_qpu.py (hunk 7 above)
Comment (co9olguy) [reply implementations/tutorial_ensemble_multi_qpu.py]: Great!

File: implementations/tutorial_ensemble_multi_qpu.py [hunk 19]
```diff
@@ -162,17 +151,29 @@ def plot_points(x_train, y_train, x_test, y_test):
...
-
-# Use qiskit.ibmq if access is available
 dev1 = qml.device("qiskit.aer", wires=4)
 devs = [dev0, dev1]
 
 ##############################################################################
+# .. note::
+#    If you have access to Rigetti hardware, you can swap out ``forest.qvm`` for ``forest.qpu``.
+#    Users with access to the IBM Q Experience can swap ``qiskit.aer`` for ``qiskit.ibmq`` and
+#    specify their chosen backend (see `here
```
Comment (co9olguy) [inline implementations/tutorial_ensemble_multi_qpu.py]: yikes the documentation for pl-qiskit looks bad! :grinning:

**Comment by co9olguy** [comment]
:ice_cream: 
Just one layout suggestion
